### PR TITLE
add deploy workflow

### PR DESCRIPTION
**Pull request by johnake**
### Context

Create a deploy workflow deployv2/action. This github workflow will be used to build and deploy the application

### Changes proposed in this pull request

IGoal: deploy to dev AKS environment automatically via Github actions. If the deployment fails, it must not impact the existing workflow.

### Guidance to review

Include any useful information needed to review this change.
Include any dependencies that are required for this change.

### Checklist

- [ ] Attach to Trello card
- [ ] Rebased main
- [ ] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

File: .github/workflows/build-and-deploy.yml [hunk 1]
```diff
@@ -60,6 +60,33 @@ jobs:
...
+    concurrency: deploy_v2_review_${{ github.event.pull_request.number }}
+    needs: [docker]
+    runs-on: ubuntu-latest
+    if: contains(github.event.pull_request.labels.*.name, 'deploy_v2')
+    environment:
+      name: review
+
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ./.github/workflows/actions/deploy-environment
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:74]: this would use the deploy_v2 action

File: .github/workflows/build-and-deploy.yml [hunk 2]
```diff
@@ -60,6 +60,33 @@ jobs:
...
             Review app deployed to ${{ steps.deploy.outputs.environment_url }}
 
+  deploy_v2_review:
+    name: Deploy to review environment
+    concurrency: deploy_v2_review_${{ github.event.pull_request.number }}
+    needs: [docker]
+    runs-on: ubuntu-latest
+    if: contains(github.event.pull_request.labels.*.name, 'deploy_v2')
+    environment:
+      name: review
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:70]: review_aks?

File: .github/workflows/build-and-deploy.yml [hunk 3]
```diff
@@ -60,6 +60,33 @@ jobs:
...
+
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ./.github/workflows/actions/deploy-environment
+        id: deploy_v2
+        with:
+          environment_name: review
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:77]: review_aks

File: Makefile [hunk 4]
```diff
@@ -78,7 +78,11 @@ review:
 	$(eval backend_config=-backend-config="key=review/review$(env).tfstate")
 	$(eval export TF_VAR_app_suffix=$(env))
 
+.PHONY: review_aks
+review_aks: aks ## Specify review aks environment
```
Comment (RMcVelia) [inline Makefile:82]: need to pass pr_id and set backend_config similar to review env above

File: .github/workflows/actions/build-and-deploy.yml
[the remark this answers is not in the document]
Comment (johnake) [reply .github/workflows/actions/build-and-deploy.yml:1]: Was meant to remove it. Removed now

File: .github/workflows/build-and-deploy.yml (hunk 1 above)
Comment (johnake) [reply .github/workflows/build-and-deploy.yml:74]: Fixed

File: .github/workflows/build-and-deploy.yml (hunk 2 above)
Comment (johnake) [reply .github/workflows/build-and-deploy.yml:70]: Fixed

File: Makefile
Comment (RMcVelia) [inline Makefile:1]: line 214 should be 
az deployment sub create -l "UK South" --template-uri "https://raw.githubusercontent.com/DFE-Digital/tra-shared-services/main/azure/resourcedeploy.json" --parameters "resourceGroupName=${AZURE_RESOURCE_PREFIX}-faltrn-${CONFIG_SHORT}-rg" 'tags=${RG_TAGS}' "tfStorageAccountName=${AZURE_RESOURCE_PREFIX}faltrntfstate${CONFIG_SHORT}" "tfStorageContainerName=faltrn-tfstate" "dbBackupStorageAccountName=false" "keyVaultName=${AZURE_RESOURCE_PREFIX}-faltrn-${CONFIG_SHORT}-kv"

line 217 would then be the same but with --what-if at the env

File: terraform/aks/workspace_variables/review_aks.backend.tfvars [hunk 5]
```diff
@@ -0,0 +1,3 @@
+resource_group_name  = "s189t01-faltrn-rv-rg"
+storage_account_name = "s189t01faltrntfstatervsa"
+key                  = "terraform.tfstate"
```
Comment (RMcVelia) [inline terraform/aks/workspace_variables/review_aks.backend.tfvars:3]: drop the key as we pass from the Makefile

File: .github/workflows/build-and-deploy.yml [hunk 6]
```diff
@@ -60,6 +60,60 @@ jobs:
...
+        with:
+          message: |
+            AKS review app deployed to ${{ steps.deploy_v2.outputs.environment_url }}
+
+  deploy_v2_dev:
+    name: Deploy to development_aks environment
+    concurrency: deploy_v2_development_${{ github.event.pull_request.number }}
+    needs: [docker]
+    runs-on: ubuntu-latest
+    if: contains(github.event.pull_request.labels.*.name, 'deploy_v2')
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:95]: perm envs not deployed on deploy_v2 label, but on merge to main
if: github.ref == 'refs/heads/main' && github.event_name == 'push'

File: .github/workflows/build-and-deploy.yml [hunk 7]
```diff
@@ -60,6 +60,60 @@ jobs:
...
+      - name: Post sticky pull request comment
+        if: github.event_name == 'pull_request'
+        uses: marocchino/sticky-pull-request-comment@v2
+        with:
+          message: |
+            AKS review app deployed to ${{ steps.deploy_v2.outputs.environment_url }}
+
+  deploy_v2_dev:
+    name: Deploy to development_aks environment
+    concurrency: deploy_v2_development_${{ github.event.pull_request.number }}
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:92]: concurrency: deploy_v2_development

File: .github/workflows/build-and-deploy.yml [hunk 8]
```diff
@@ -60,6 +60,60 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ./.github/workflows/actions/deploy_v2
+        id: deploy_v2
+        with:
+          environment: development_aks
+          docker_image: ${{ needs.docker.outputs.docker_image }}
+          azure-credentials: ${{ secrets.AZURE_CREDENTIALS }}
+          arm-access-key: ${{ secrets.ARM_ACCESS_KEY }}
+          pr-id: ${{ github.event.pull_request.number }}
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:108]: remove pr-id as only valid for review apps

File: .github/workflows/build-and-deploy.yml [hunk 9]
```diff
@@ -60,6 +60,60 @@ jobs:
...
+
+      - name: Post sticky pull request comment
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:110]: we don't need this step for dev env

File: .github/workflows/build-and-deploy.yml [hunk 10]
```diff
@@ -60,6 +60,60 @@ jobs:
...
+        if: github.event_name == 'pull_request'
+        uses: marocchino/sticky-pull-request-comment@v2
+        with:
+          message: |
+            AKS dev app deployed to ${{ steps.deploy_v2.outputs.environment_url }}
+
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:116]: will need to add the smoke test step as per deploy_nonprod for paas.
will need an aks copy of the smoke-test action i.e. smoke-test-v2

File: .github/workflows/build-and-deploy.yml [hunk 11]
```diff
@@ -60,6 +60,57 @@ jobs:
...
+        uses: marocchino/sticky-pull-request-comment@v2
+        with:
+          message: |
+            AKS review app deployed to ${{ steps.deploy_v2.outputs.environment_url }}
+
+  deploy_v2_dev:
+    name: Deploy to development_aks environment
+    concurrency: deploy_v2_development
+    needs: [docker]
+    runs-on: ubuntu-latest
```
Comment (RMcVelia) [inline .github/workflows/build-and-deploy.yml:94]: add 'continue-on-error: true' as we don't want an aks dev deploy failure to stop paas deploys

**Comment by johnake** [comment]
[Add deploy workflow](https://trello.com/c/f7V6d9Dc/528-add-deploy-workflow)

**Pull request merged**